oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java

```diff
@@ -306,4 +312,23 @@ private SessionInformation getSessionInformation(Authentication principal) {
 		return sessionInformation;
 	}
 
+	private void invalidateAuthorizationTokens(OAuth2Authorization authorization) {
```

Comment:
Can you try to integrate this logic into `OAuth2AuthenticationProviderUtils.invalidate()` so we can isolate the logic in one spot?

Reply:
It is difficult to merge the added logic with the existing one, since the existing "invalidate()" method invalidates a single token, typically an authorization code or a refresh token, and sometimes invalidates access tokens or authorization codes in addition to the refresh token.

Our requirement is to invalidate both access tokens and refresh tokens. Therefore, integrating the added logic with the existing "invalidate()" method is not feasible. Instead, we created a new method called "invalidateAuthorizationTokens()" that invalidates two types of tokens.

Reply:
> Our requirement is to invalidate both access tokens and refresh tokens. Therefore, integrating the added logic with the existing "invalidate()" method is not feasible.

The addition of the following should work?

```java
static <T extends OAuth2Token> OAuth2Authorization invalidate(
		OAuth2Authorization authorization, T token) {

	OAuth2Authorization.Builder authorizationBuilder = OAuth2Authorization.from(authorization)
			.token(token,
					(metadata) ->
							metadata.put(OAuth2Authorization.Token.INVALIDATED_METADATA_NAME, true));

	if (OAuth2AuthorizationCode.class.isAssignableFrom(token.getClass())) {
		// TODO Invalidate access token (if available)


		// TODO Invalidate refresh token (if available)

	}

	return authorizationBuilder.build();
}
```

Reply:
It might work, but modifying invalidate() could cause it to lose its original functionality of invalidating a single token
For example, there may be code where invalidate() should only invalidate one token without affecting any other tokens in the sequence.

https://github.com/spring-projects/spring-authorization-server/blob/13a61034eda67f74a2162c5edbaec5243218341e/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java#L254

Reply:
What do you think about my opinion regarding the requested modification for invalidate()?

Reply:
@neochae1 

> modifying invalidate() could cause it to lose its original functionality of invalidating a single token

There are existing test cases that would fail if the current logic was changed incorrectly. I believe enhancing the logic will work for this additional use case. Please give it a try. My preference is to keep the logic in one spot.

Reply:
I have merged the logic for invalidating tokens into the existing invalidate() function. 
Also, I have relocated the logic for invalidating the code to the builder creation process for success scenario.

Reply:
@jgrandja 
Please provide additional review feedback regarding the content you revised based on the method you proposed.